File: core/src/main/java/com/graphhopper/routing/util/parsers/ModeAccessParser.java
package com.graphhopper.routing.util.parsers;

import com.graphhopper.reader.ReaderWay;
import com.graphhopper.routing.ev.BooleanEncodedValue;
import com.graphhopper.routing.ev.EdgeIntAccess;
import com.graphhopper.routing.util.FerrySpeedCalculator;
import com.graphhopper.storage.IntsRef;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.graphhopper.routing.util.parsers.OSMTemporalAccessParser.hasPermissiveTemporalRestriction;

public class ModeAccessParser implements TagParser {

    private static final Set<String> CAR_BARRIERS = Set.of("kissing_gate", "fence",
            "bollard", "stile", "turnstile", "cycle_barrier", "motorcycle_barrier", "block",
            "bus_trap", "sump_buster", "jersey_barrier");
    private static final Set<String> INTENDED = Set.of("yes", "designated", "official", "permissive", "private", "permit", "destination");
    static final Map<String, Map<String, String>> HIGHWAY_TYPE_DEFAULTS;
    static final Map<String, Map<String, String>> BARRIER_TYPE_DEFAULTS;

    // default implied access for unknown highway types
    private static final Map<String, String> UNKNOWN_HIGHWAY_DEFAULT = Map.of("access", "no");

    static {
        Map<String, Map<String, String>> m = new HashMap<>();
        // routable highway types with no special access defaults (empty map = fully accessible)
        for (String hw : List.of("motorway", "motorway_link", "trunk", "trunk_link",
                "primary", "primary_link", "secondary", "secondary_link",
                "tertiary", "tertiary_link", "unclassified", "residential",
                "living_street", "service", "road", "track"))
            m.put(hw, Map.of());
        // routable highway types with specific access defaults
        m.put("steps", Map.of("access", "no"));
        m.put("footway", Map.of("motor_vehicle", "no"));
        m.put("cycleway", Map.of("motor_vehicle", "no"));
        m.put("pedestrian", Map.of("motor_vehicle", "no"));
        m.put("path", Map.of("motor_vehicle", "no"));
        m.put("bridleway", Map.of("motor_vehicle", "no"));
        m.put("busway", Map.of("access", "no", "bus", "designated"));
        HIGHWAY_TYPE_DEFAULTS = Map.copyOf(m);

        Map<String, Map<String, String>> b = new HashMap<>();
        b.put("bus_trap", Map.of("motor_vehicle", "no", "bus", "yes"));
        b.put("sump_buster", Map.of("motor_vehicle", "no", "bus", "yes"));
        BARRIER_TYPE_DEFAULTS = Map.copyOf(b);
    }
    private static final Set<String> ONEWAYS_FW = Set.of("yes", "true", "1");
    private final Set<String> restrictedValues;
    private final List<String> restrictionKeys;
    private final List<String> vehicleForward;
    private final List<String> vehicleBackward;
    private final List<String> ignoreOnewayKeys;
    private final BooleanEncodedValue accessEnc;
    private final BooleanEncodedValue roundaboutEnc;
    private final boolean skipEmergency;
    private final Set<String> barriers;

    public ModeAccessParser(List<String> restrictionKeys, BooleanEncodedValue accessEnc,
                            boolean skipEmergency, BooleanEncodedValue roundaboutEnc,
                            Set<String> restrictions, Set<String> barriers) {
        this.accessEnc = accessEnc;
        this.roundaboutEnc = roundaboutEnc;
        this.restrictionKeys = restrictionKeys;
        vehicleForward = restrictionKeys.stream().map(r -> r + ":forward").toList();
        vehicleBackward = restrictionKeys.stream().map(r -> r + ":backward").toList();
        ignoreOnewayKeys = restrictionKeys.stream().map(r -> "oneway:" + r).toList();
        restrictedValues = restrictions.isEmpty() ? Set.of("no", "restricted", "military", "emergency") : restrictions;
        this.barriers = barriers.isEmpty() ? CAR_BARRIERS : barriers;
        if (restrictedValues.contains(""))
            throw new IllegalArgumentException("restriction values cannot contain empty string");
        this.skipEmergency = skipEmergency;
    }

    @Override
    public void handleWayTags(int edgeId, EdgeIntAccess edgeIntAccess, ReaderWay way, IntsRef relationFlags) {
        String highwayValue = way.getTag("highway");
        if (highwayValue == null && !FerrySpeedCalculator.isFerry(way))
            return;
        if (skipEmergency && "service".equals(highwayValue) && "emergency_access".equals(way.getTag("service")))
            return;

        Map<String, String> defaults = highwayValue == null ? Map.of() : HIGHWAY_TYPE_DEFAULTS.getOrDefault(highwayValue, UNKNOWN_HIGHWAY_DEFAULT);
        int firstIndex = -1;
        String firstValue = "";
        for (int i = 0; i < restrictionKeys.size(); i++) {
            String key = restrictionKeys.get(i);
            String explicit = way.getTag(key);
            if (explicit != null) {
                firstIndex = i;
                firstValue = explicit;
                break;
            }
            String implied = defaults.get(key);
            if (implied != null) {
                firstIndex = i;
                firstValue = implied;
                break;
            }
        }
        if (isRestricted(firstValue) && !hasPermissiveTemporalRestriction(way, firstIndex, restrictionKeys, INTENDED))
            return;

        if (way.hasTag("gh:barrier_edge") && way.hasTag("node_tags")) {
            List<Map<String, Object>> nodeTags = way.getTag("node_tags", null);
            Map<String, Object> firstNodeTags = nodeTags.get(0);
            String barrierValue = firstNodeTags.containsKey("barrier") ? (String) firstNodeTags.get("barrier") : "";
            Map<String, String> barrierDefaults = BARRIER_TYPE_DEFAULTS.getOrDefault(barrierValue, Map.of());
            // Walk restriction keys checking explicit node tags and barrier type defaults
            String nodeValue = "";
            for (String key : restrictionKeys) {
                String explicit = (String) firstNodeTags.get(key);
                if (explicit != null) {
                    nodeValue = explicit;
                    break;
                }
                String implied = barrierDefaults.get(key);
                if (implied != null) {
                    nodeValue = implied;
                    break;
                }
            }
            if (restrictedValues.contains(nodeValue))
                return;
            if ("yes".equals(firstNodeTags.get("locked")) && !INTENDED.contains(nodeValue))
                return;
            if (!INTENDED.contains(nodeValue) && barriers.contains(barrierValue))
                return;
        }

        if (FerrySpeedCalculator.isFerry(way)) {
            boolean isCar = restrictionKeys.contains("motorcar");
            if (INTENDED.contains(firstValue)
                    // implied default is allowed only if foot and bicycle is not specified:
                    || isCar && firstValue.isEmpty() && !way.hasTag("foot") && !way.hasTag("bicycle")
                    // if hgv is allowed then smaller trucks and cars are allowed too even if not specified
                    || isCar && way.hasTag("hgv", "yes")) {
                // ferry is allowed via explicit tag
            } else {
                return;
            }
        }

        boolean isRoundabout = roundaboutEnc.getBool(false, edgeId, edgeIntAccess);
        boolean ignoreOneway = "no".equals(way.getFirstValue(ignoreOnewayKeys));
        boolean isBwd = isBackwardOneway(way);
        if (!ignoreOneway && (isBwd || isRoundabout || isForwardOneway(way))) {
            accessEnc.setBool(isBwd, edgeId, edgeIntAccess, true);
        } else {
            accessEnc.setBool(false, edgeId, edgeIntAccess, true);
            accessEnc.setBool(true, edgeId, edgeIntAccess, true);
        }

    }

    protected boolean isBackwardOneway(ReaderWay way) {
        // vehicle:forward=no is like oneway=-1
        return way.hasTag("oneway", "-1") || "no".equals(way.getFirstValue(vehicleForward));
    }

    protected boolean isForwardOneway(ReaderWay way) {
        // vehicle:backward=no is like oneway=yes
        return way.hasTag("oneway", ONEWAYS_FW) || "no".equals(way.getFirstValue(vehicleBackward));
    }

    /**
     * Handles semicolon-separated restriction values like "agricultural;forestry".
     * Returns true (restricted) only if no part is an INTENDED value, and at least one part is restricted.
     */
    private boolean isRestricted(String value) {
        if (value.isEmpty()) return false;
        if (!value.contains(";")) return restrictedValues.contains(value);
        String[] parts = value.split(";");
        for (String part : parts) {
            if (INTENDED.contains(part)) return false;
        }
        for (String part : parts) {
            if (restrictedValues.contains(part)) return true;
        }
        return false;
    }
}

File: core/src/main/kotlin/com/graphhopper/routing/util/parsers/ModeAccessParser.kt
package com.graphhopper.routing.util.parsers

import com.graphhopper.reader.ReaderWay
import com.graphhopper.routing.ev.BooleanEncodedValue
import com.graphhopper.routing.ev.EdgeIntAccess
import com.graphhopper.routing.util.FerrySpeedCalculator
import com.graphhopper.routing.util.parsers.OSMTemporalAccessParser.Companion.hasPermissiveTemporalRestriction
import com.graphhopper.storage.IntsRef

class ModeAccessParser @JvmOverloads constructor(
    private val restrictionKeys: List<String>,
    private val accessEnc: BooleanEncodedValue,
    private val skipEmergency: Boolean,
    private val roundaboutEnc: BooleanEncodedValue,
    allow: Set<String> = emptySet(),
    restrict: Set<String> = emptySet()
) : TagParser {

    private val intended: MutableSet<String> = HashSet(INTENDED)
    private val restricted: MutableSet<String> = HashSet(RESTRICTED)
    private val vehicleForward: List<String> = restrictionKeys.map { "$it:forward" }
    private val vehicleBackward: List<String> = restrictionKeys.map { "$it:backward" }
    private val onewayModeKeys: List<String> = restrictionKeys.map { "oneway:$it" }

    init {
        for (value in allow) {
            if (restricted.remove(value))
                intended.add(value)
            else if (!intended.contains(value))
                throw IllegalArgumentException("cannot allow '$value' — not a known restricted value")
        }
        for (value in restrict) {
            if (intended.contains(value))
                throw IllegalArgumentException("cannot restrict '$value' — it is an intended value")
            restricted.add(value)
        }
    }

    override fun handleWayTags(edgeId: Int, edgeIntAccess: EdgeIntAccess, way: ReaderWay, relationFlags: IntsRef?) {
        val highwayValue = way.getTag("highway")
        if (highwayValue != null || FerrySpeedCalculator.isFerry(way)) {
            handleHighwayAndFerryTags(edgeId, edgeIntAccess, way, highwayValue)
        }
        // don't want platforms and other random stuff here for now
    }

    private fun handleHighwayAndFerryTags(edgeId: Int, edgeIntAccess: EdgeIntAccess, way: ReaderWay, highwayValue: String?) {
        if (skipEmergency && "service" == highwayValue && "emergency_access" == way.getTag("service"))
            return

        var defaults: Map<String, String> =
            if (highwayValue == null) emptyMap() else HIGHWAY_TYPE_DEFAULTS.getOrDefault(highwayValue, emptyMap())
        // motorroad=yes is an annoying special case: it's not a highway=* value but a separate tag
        // that implies foot=no, bicycle=no. If we find more tags like this, we'll need a more
        // general mechanism for non-highway implied defaults.
        if (way.hasTag("motorroad", "yes")) {
            val merged = HashMap(defaults)
            MOTORROAD_DEFAULTS.forEach { (k, v) -> merged.putIfAbsent(k, v) }
            defaults = merged
        }
        var firstIndex = -1
        var firstValue = ""
        for (i in restrictionKeys.indices) {
            val key = restrictionKeys[i]
            val explicit = way.getTag(key)
            if (explicit != null) {
                if (intended.contains(explicit) || restricted.contains(explicit)) {
                    firstIndex = i
                    firstValue = explicit
                    break
                }
                // unknown value — fall through to implied default for same key, then keep looking
            }
            val implied = defaults[key]
            if (implied != null) {
                firstIndex = i
                firstValue = implied
                break
            }
        }
        if (restricted.contains(firstValue) && !hasPermissiveTemporalRestriction(way, firstIndex, restrictionKeys, intended))
            return

        if (way.hasTag("gh:barrier_edge") && way.hasTag("node_tags")) {
            val nodeTags: List<Map<String, Any>> = way.getTag<List<Map<String, Any>>?>("node_tags", null)!!
            val firstNodeTags = nodeTags[0]
            val barrierValue = if (firstNodeTags.containsKey("barrier")) firstNodeTags["barrier"] as String else ""
            val barrierDefaults = BARRIER_TYPE_DEFAULTS.getOrDefault(barrierValue, emptyMap())
            // Walk restriction keys checking explicit node tags and barrier type defaults
            var nodeValue = ""
            for (key in restrictionKeys) {
                val explicit = firstNodeTags[key] as String?
                if (explicit != null) {
                    if (intended.contains(explicit) || restricted.contains(explicit)) {
                        nodeValue = explicit
                        break
                    }
                }
                val implied = barrierDefaults[key]
                if (implied != null) {
                    nodeValue = implied
                    break
                }
            }
            if (restricted.contains(nodeValue))
                return
            if ("yes" == firstNodeTags["locked"] && !intended.contains(nodeValue))
                return
        }

        if (FerrySpeedCalculator.isFerry(way)) {
            val isCar = restrictionKeys.contains("motorcar")
            if (intended.contains(firstValue)
                // implied default is allowed only if foot and bicycle is not specified:
                || isCar && firstValue.isEmpty() && !way.hasTag("foot") && !way.hasTag("bicycle")
                // if hgv is allowed then smaller trucks and cars are allowed too even if not specified
                || isCar && way.hasTag("hgv", "yes")
            ) {
                // ferry is allowed via explicit tag
            } else {
                return
            }
        }

        val isRoundabout = roundaboutEnc.getBool(false, edgeId, edgeIntAccess)
        if (isForwardAccessible(way))
            accessEnc.setBool(false, edgeId, edgeIntAccess, true)
        if (isBackwardAccessible(way, isRoundabout))
            accessEnc.setBool(true, edgeId, edgeIntAccess, true)
    }

    /**
     * The two directions are computed independently so that blockings from different
     * tag families stack: e.g. oneway=yes closing backward plus bus:forward=no closing
     * forward leaves a bus with nowhere to go, rather than being treated as a single
     * backward-oneway by the dominant rule.
     *
     * Within each direction, rules are evaluated top-to-bottom; the first rule that
     * matches wins. More-specific tags come first, so a mode-specific tag always beats
     * a generic one.
     */
    private fun isForwardAccessible(way: ReaderWay): Boolean {
        val modeOneway = way.getFirstValue(onewayModeKeys)
        // 1. explicit per-mode oneway override: oneway:<mode>=yes/-1/no
        if ("no" == modeOneway) return true
        if (ONEWAYS_FW.contains(modeOneway)) return true
        if ("-1" == modeOneway) return false
        // 2. mode-specific directional prohibition: <mode>:forward=no
        if ("no" == way.getFirstValue(vehicleForward)) return false
        // 3. generic oneway tag (forward direction is only closed by oneway=-1;
        //    motorway/roundabout imply forward-oneway, which leaves forward open)
        return !way.hasTag("oneway", "-1")
    }

    private fun isBackwardAccessible(way: ReaderWay, isRoundabout: Boolean): Boolean {
        val modeOneway = way.getFirstValue(onewayModeKeys)
        // 1. explicit per-mode oneway override
        if ("no" == modeOneway) return true
        if (ONEWAYS_FW.contains(modeOneway)) return false
        if ("-1" == modeOneway) return true
        // 2. mode-specific directional prohibition: <mode>:backward=no
        if ("no" == way.getFirstValue(vehicleBackward)) return false
        // 3. generic oneway tag, then implied oneway from highway type / roundabout.
        //    oneway=no explicitly relaxes the motorway/roundabout implication.
        if (way.hasTag("oneway", ONEWAYS_FW)) return false
        if (way.hasTag("oneway", "no")) return true
        if (way.hasTag("highway", "motorway", "motorway_link")) return false
        return !isRoundabout
    }

    companion object {
        private val INTENDED = setOf("yes", "designated", "official", "permissive", "destination")
        private val RESTRICTED = setOf(
            "no", "restricted", "military", "emergency",
            "private", "permit", "service", "delivery", "customers", "agricultural", "forestry"
        )
        private val MOTORROAD_DEFAULTS = mapOf("foot" to "no", "bicycle" to "no")
        private val ONEWAYS_FW = setOf("yes", "true", "1")

        // https://wiki.openstreetmap.org/wiki/OSM_tags_for_routing/Access_restrictions
        // For implied bicycle defaults we use "dismount" rather than "no" on highways where
        // pushing a bike is physically reasonable (pedestrian streets, footways, stairs, indoor
        // corridors, station platforms). Our routing layer treats DISMOUNT as "can push" —
        // usable but qualified — whereas NO is treated as a hard block. motorway/motorway_link
        // keep "no" because bikes are legally prohibited even on foot. Explicit bicycle=no tags
        // on a way still resolve to NO; only the implicit highway-type defaults become DISMOUNT.
        internal val HIGHWAY_TYPE_DEFAULTS: Map<String, Map<String, String>> = mapOf(
            "motorway" to mapOf("motor_vehicle" to "designated", "foot" to "no", "bicycle" to "no"),
            "motorway_link" to mapOf("motor_vehicle" to "designated", "foot" to "no", "bicycle" to "no"),
            "steps" to mapOf("motor_vehicle" to "no", "bicycle" to "dismount", "foot" to "designated"),
            "footway" to mapOf("motor_vehicle" to "no", "bicycle" to "dismount", "foot" to "designated"),
            "cycleway" to mapOf("motor_vehicle" to "no", "bicycle" to "designated", "foot" to "no"),
            "pedestrian" to mapOf("motor_vehicle" to "no", "bicycle" to "dismount", "foot" to "designated"),
            "path" to mapOf("motor_vehicle" to "no", "foot" to "yes", "bicycle" to "yes"),
            "bridleway" to mapOf("motor_vehicle" to "no", "foot" to "yes", "bicycle" to "yes"),
            "busway" to mapOf("access" to "no", "bus" to "designated"),
            "construction" to mapOf("access" to "no"),
            "proposed" to mapOf("access" to "no"),
            "raceway" to mapOf("access" to "no"),
            // corridor stays bicycle=no (not "dismount") because BikeCommonAccessParser doesn't list
            // it in allowedHighways — bike_access is already false there. Marking road_access as
            // DISMOUNT would suggest "can push" while access says "can't use it at all".
            "corridor" to mapOf("motor_vehicle" to "no", "bicycle" to "no", "foot" to "yes"),
            "platform" to mapOf("motor_vehicle" to "no", "bicycle" to "dismount", "foot" to "yes")
        )

        // https://wiki.openstreetmap.org/wiki/Key:barrier
        internal val BARRIER_TYPE_DEFAULTS: Map<String, Map<String, String>> = mapOf(
            "fence" to mapOf("access" to "no"),
            "wall" to mapOf("access" to "no"),
            "hedge" to mapOf("access" to "no"),
            "retaining_wall" to mapOf("access" to "no"),
            "city_wall" to mapOf("access" to "no"),
            "ditch" to mapOf("access" to "no"),
            "kerb" to mapOf("vehicle" to "yes", "foot" to "yes"),
            "cattle_grid" to mapOf("motor_vehicle" to "yes", "foot" to "yes", "bicycle" to "yes"),
            "bollard" to mapOf("motor_vehicle" to "no", "foot" to "yes", "bicycle" to "yes"),
            "block" to mapOf("motor_vehicle" to "no", "foot" to "yes", "bicycle" to "yes"),
            "log" to mapOf("motor_vehicle" to "no", "foot" to "yes", "bicycle" to "yes"),
            "chain" to mapOf("motor_vehicle" to "no", "foot" to "yes", "bicycle" to "yes"),
            "jersey_barrier" to mapOf("motor_vehicle" to "no", "foot" to "yes", "bicycle" to "yes"),
            "cycle_barrier" to mapOf("motor_vehicle" to "no", "foot" to "yes", "bicycle" to "yes"),
            "motorcycle_barrier" to mapOf("motor_vehicle" to "no", "foot" to "yes", "bicycle" to "yes"),
            "bus_trap" to mapOf("motor_vehicle" to "no", "bus" to "yes", "foot" to "yes", "bicycle" to "yes"),
            "sump_buster" to mapOf("motor_vehicle" to "no", "bus" to "yes", "foot" to "yes", "bicycle" to "yes"),
            "kissing_gate" to mapOf("vehicle" to "no", "foot" to "yes"),
            "stile" to mapOf("vehicle" to "no", "foot" to "yes"),
            "turnstile" to mapOf("vehicle" to "no", "foot" to "yes")
        )
    }
}
